### ISSUE TITLE
Move to new scene data model

### ISSUE THREAD

**Issue by robertlong** [MozillaReality/Spoke#349]
## Background
Spoke was initially designed with a entity-component data model. The scene was defined as a series of entities (Object3D's) that had components (data stored in object3D.userData). We intended to expose this directly to the user by letting them compose entities out of components.

<img width="720" alt="screen shot 2018-10-19 at 10 32 15 am" src="https://user-images.githubusercontent.com/1753624/47234346-63e99480-d38a-11e8-8fb8-2d2072656347.png">

We built out most of the code necessary for this to work, however the code was brittle and complex. ThreeJS is not backed by an entity-component model and mapping to one is quite difficult. We ended up having to hack around the library in order to get a pattern we liked and the end result wasn't very usable. Also, the complexity presented to the user with this entity component model made Spoke hard to learn. So we back to the drawing board.

Today, Spoke exposes a variety of nodes for you to compose scenes with. They are still backed by this entity-component model, but we hide that complexity from the user. The remaining entity-component code is lying around not doing too much and it's design makes it difficult to build the correct UX around these different nodes. Visual helpers (lights and spawn points), file loading (glTF model), and hiding components (Floor Plan) are not implemented in a very clean way.

## Proposal
I propose that we move to a scene data model that better fits our UI, different node types, and ThreeJS's scene graph. I believe that we can accomplish this by simply adopting an inheritance based scene data model and creating our own super classes of the different Object3D types. Each object will be registered with an associated React component that defines the editor for the selected object type. It will receive the selected object and the editor as props. The object can then define the exact look of the property panel as well as any custom behavior needed to create the correct UX for that object.

For example the Model node. The model node has a file property that needs to asynchronously load it's subgraph from the `THREE.GLTFLoader`. It's subgraph should not be serialized when saving as a `.spoke` file, but should be exported when exporting as a glTF. When serializing as a `.spoke` file it should save all of its props but shouldn't when exporting as a glTF. When you change the file property it should show the file dialog, allow you to select a new model, and then load that model with the `THREE.GLTFLoader`. It also needs to show you the available animation clips when that model has been loaded.

All of the scene graph state associated with this object should be stored in the object3D itself. The UI state should be stored in that object's react component. Custom logic for serialization, deserialization, exporting, copying, adding, and removing should be included in that object3D.

### Registering Node Types
```js
editor.registerNodeType(displayName, iconCSSClass, object3DConstructor, editorReactComponent)
editor.registerNodeType("Model", "model-icon-class", ModelNode, ModelNodeEditor)
```

### Object3D Class
```js
class ModelNode extends THREE.Object3D {
  constructor() {
    super();
    this.type = "ModelNode";
    this.isNode = true; // used for determining selection root etc.
    this.model = null;
    this.src = null;
  }

  async setModel(src) {
    if (this.model) {
      this.remove(model);
    }

    const model = await gltfCache.get(src);
    this.add(model);
    this.src = src;
  }

  copy(target) {
    // Exclude this.model, copy props, etc.
  }

  // Only object3Ds in the scene that implement this method will be serialized.
  serialize() {
    // Return the props you want to be saved into the .spoke file
    return { src: this.src };
  }

  // Use the props from the .spoke file to load the node.
  // Async so that we can wait for the node to be fully loaded.
  async deserialize(json) {
    await this.setModel(json.src);
  }

  prepareForExport() {
    // Mutate this object so that it is ready to be exported by THREE.GLTFExporter.
    // You can remove this node by calling this.parent.remove(this) if you don't want this node to be exported.
    // You can also remove any helper objects here, modify userData etc.
  }
}
```

### React Component
```js
class ModelNodeEditor {
  onChangeSrc = (src) {
    // We could also await this if we wanted to show a progress dialog.
    this.props.object3D.setModel(src);
  }

  render() {
    // Editor is not passed by default but you can connect an editor context if needed.
    const { displayName, iconCSSClass, object3D } = this.props;
    return (
      <PropertyGroup name={displayName}>
        <InputGroup name="src">
          <FileInput value={object3D.src} onChange={this.onChangeSrc}/>
        </InputGroup>
      </PropertyGroup>
    );
  }
}
```

This implementation is rough and there are a lot of pieces that I am unsure about. I'd appreciate any feedback on this design or any alternatives that make Spoke more robust and easier to build out new node types.

**Comment by robertlong** [comment]
Here's another pass at this concept. I really like the separation of concerns between the editor specific Object3D methods and the "runtime" Object3D. An example of a "runtime" object 3D that we use currently could be `THREE.Sky`, `THREE.Group`, etc. Being able to mixin the editor functionality allows us to bring in third party components and define how they should be serialized, deserialized and exported.

```javascript
import React, { Component } from "react";
import PropTypes from "prop-types";
import PropertyGroup from "../../ui/PropertyGroup";
import InputGroup from "../../ui/InputGroup";
import FileInput from "../../ui/inputs/FileInput";
import THREE from "../three";
import NodeMixin from "./NodeMixin";

class ModelNodeEditor extends Component {
  static propTypes = {
    object3D: PropTypes.object
  };

  onChangeSrc = src => {
    this.props.object3D.setModel(src);
  };

  render() {
    const { object3D } = this.props;

    return (
      <PropertyGroup name="Model">
        <InputGroup name="src">
          <FileInput value={object3D.src} onChange={this.onChangeSrc} />
        </InputGroup>
      </PropertyGroup>
    );
  }
}

// Runtime Object3D
class ModelNode extends THREE.Object3D {
  constructor() {
    super();
    this.type = "ModelNode";
    this.src = null;
    this.model = null;
  }
}

// Editor only Object3D
export default class EditorModelNode extends NodeMixin(ModelNode) {
  static editorComponent = ModelNodeEditor;

  static async deserialize(json) {
    const node = new ModelNode();
    await node.setModel(json.src);
    return node;
  }

  serialize() {
    return { src: this.src };
  }

  async setModel(src) {
    const model = await this.editor.loadGLTF(src);
    this.model = model;
    this.src = src;
  }

  prepareForExport() {
    this.userData.gltfExtensions = {
      MOZ_hubs_components: {
        gltfModel: {
          src: this.src
        }
      }
    };
  }
}
```

NodeMixin.js
```javascript
export default Object3DClass =>
  class extends Object3DClass {
    constructor(...args) {
      super(...args);
      // Used for determining what objects to select when clicking in the viewport (Replaces userData._selectionRoot)
      // Also used for determining what objects to serialize when saving .spoke file.
      this.isNode = true;
    }
  };
```

**Comment by robertlong** [comment]
This is shipped in v0.6.0
